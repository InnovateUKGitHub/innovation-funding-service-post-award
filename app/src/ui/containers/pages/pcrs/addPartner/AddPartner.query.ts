import { graphql } from "react-relay";

export const addPartnerWorkflowQuery = graphql`
  query AddPartnerWorkflowQuery($projectId: ID!, $pcrItemId: ID!) {
    salesforce {
      uiapi {
        query {
          Acc_ProjectParticipant__c(
            where: { Acc_ProjectId__c: { eq: $projectId } }
            orderBy: { Acc_AccountId__r: { Name: { order: ASC } } }
            first: 2000
          ) {
            edges {
              node {
                Id
                Acc_AccountId__r {
                  Name {
                    value
                  }
                }
                Acc_OrganisationType__c {
                  value
                }
                Acc_ParticipantStatus__c {
                  value
                }
                Acc_ProjectRole__c {
                  value
                }
              }
            }
          }
          Acc_ProjectChangeRequest__c(where: { Id: { eq: $pcrItemId } }, first: 1) {
            edges {
              node {
                Id
                Acc_ParticipantSize__c {
                  value
                }
                Acc_ProjectRole__c {
                  value
                }
                Acc_ParticipantType__c {
                  value
                }
                Acc_OrganisationName__c {
                  value
                }
                Acc_CommercialWork__c {
                  value
                }
                Acc_Location__c {
                  value
                }
                Acc_ProjectPostcode__c {
                  value
                }
                Acc_ProjectCity__c {
                  value
                }
                RecordType {
                  Name {
                    value
                    label
                  }
                  DeveloperName {
                    value
                  }
                }
                ContentDocumentLinks(first: 2000, orderBy: { ContentDocument: { CreatedDate: { order: DESC } } }) {
                  edges {
                    node {
                      Id
                      LinkedEntityId {
                        value
                      }
                      isFeedAttachment
                      isOwner
                      ContentDocument {
                        Id
                        LastModifiedBy {
                          ContactId {
                            value
                          }
                        }
                        Description {
                          value
                        }
                        CreatedDate {
                          value
                        }
                        LatestPublishedVersionId {
                          value
                        }
                        FileExtension {
                          value
                        }
                        Title {
                          value
                        }
                        ContentSize {
                          value
                        }
                        CreatedBy {
                          Name {
                            value
                          }
                          Id
                        }
                      }
                    }
                  }
                }
              }
            }
          }
          Acc_Project__c(where: { Id: { eq: $projectId } }, first: 1) {
            edges {
              node {
                Id
                roles {
                  isPm
                  isFc
                  isMo
                }
                Acc_CompetitionType__c {
                  value
                }
              }
            }
          }
        }
      }
    }
  }
`;
