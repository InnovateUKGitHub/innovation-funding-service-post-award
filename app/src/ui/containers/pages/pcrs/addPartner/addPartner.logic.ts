import { useLazyLoadQuery } from "react-relay";
import { addPartnerWorkflowQuery } from "./AddPartner.query";
import { AddPartnerWorkflowQuery } from "./__generated__/AddPartnerWorkflowQuery.graphql";
import { getFirstEdge } from "@gql/selectors/edges";
import { mapToProjectDto } from "@gql/dtoMapper/mapProjectDto";
import { mapPcrItemDto } from "@gql/dtoMapper/mapPcrDto";
import { mapToPartnerDtoArray } from "@gql/dtoMapper/mapPartnerDto";
import { sortPartnersLeadFirst } from "@framework/util/partnerHelper";
import { mapToDocumentSummaryDto } from "@gql/dtoMapper/mapDocumentsDto";

export const useAddPartnerWorkflowQuery = (projectId: ProjectId, pcrItemId: PcrItemId, fetchKey: number) => {
  const data = useLazyLoadQuery<AddPartnerWorkflowQuery>(
    addPartnerWorkflowQuery,
    {
      projectId,
      pcrItemId,
    },
    {
      fetchPolicy: "network-only",
      fetchKey,
    },
  );

  const { node: projectNode } = getFirstEdge(data?.salesforce?.uiapi?.query?.Acc_Project__c?.edges);
  const { node: pcrNode } = getFirstEdge(data?.salesforce?.uiapi?.query?.Acc_ProjectChangeRequest__c?.edges);
  const project = mapToProjectDto(projectNode, ["projectNumber", "status", "title", "roles", "competitionType"]);

  const partners = sortPartnersLeadFirst(
    mapToPartnerDtoArray(
      data?.salesforce?.uiapi?.query?.Acc_ProjectParticipant__c?.edges ?? [],
      ["id", "isLead", "isWithdrawn", "name"],
      {},
    ),
  );

  const documents = (pcrNode?.ContentDocumentLinks?.edges ?? []).map(node =>
    mapToDocumentSummaryDto(
      node,
      ["id", "dateCreated", "description", "fileName", "fileSize", "isOwner", "uploadedBy", "link", "linkedEntityId"],
      {
        type: "pcr",
        projectId,
        pcrId: pcrItemId,
      },
    ),
  );

  const pcrItem = mapPcrItemDto(
    pcrNode,
    [
      "projectRole",
      "partnerType",
      "isCommercialWork",
      "organisationType",
      "organisationName",
      "participantSize",
      "projectCity",
      "projectLocation",
      "projectPostcode",
    ],
    {},
  );

  return { project, pcrItem, partners, documents, fragmentRef: data?.salesforce?.uiapi };
};
