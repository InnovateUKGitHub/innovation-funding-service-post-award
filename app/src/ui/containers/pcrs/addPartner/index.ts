export * from "./academicOrganisationStep";
export * from "./addPartnerSummary";
export * from "./financeContactStep";
export * from "./financeDetailsStep";
export * from "./organisationDetailsSteps";
export * from "./projectLocationStep";
export * from "./roleAndOrganisationStep";
