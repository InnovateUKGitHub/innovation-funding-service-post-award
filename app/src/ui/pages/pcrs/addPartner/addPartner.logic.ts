import { useLazyLoadQuery } from "react-relay";
import { addPartnerWorkflowQuery } from "./AddPartner.query";
import { AddPartnerWorkflowQuery } from "./__generated__/AddPartnerWorkflowQuery.graphql";
import { getFirstEdge } from "@gql/selectors/edges";
import { mapToProjectDto } from "@gql/dtoMapper/mapProjectDto";
import { mapPcrItemDto } from "@gql/dtoMapper/mapPcrDto";
import { mapToPartnerDtoArray } from "@gql/dtoMapper/mapPartnerDto";
import { sortPartnersLeadFirst } from "@framework/util/partnerHelper";
import { mapToDocumentSummaryDto } from "@gql/dtoMapper/mapDocumentsDto";
import { mapToCostCategoryDtoArray } from "@gql/dtoMapper/mapCostCategoryDto";
import { mapPcrSpendProfileArray } from "@gql/dtoMapper/mapPcrSpendProfile";
import { PCROrganisationType } from "@framework/constants/pcrConstants";
import { CostCategoryType } from "@framework/constants/enums";
import { PcrSpendProfileDto } from "@framework/dtos/pcrSpendProfileDto";
import { CostCategoryDto } from "@framework/dtos/costCategoryDto";

export const useAddPartnerWorkflowQuery = (projectId: ProjectId, pcrItemId: PcrItemId, fetchKey: number) => {
  const data = useLazyLoadQuery<AddPartnerWorkflowQuery>(
    addPartnerWorkflowQuery,
    {
      projectId,
      pcrItemId,
    },
    {
      fetchPolicy: "network-only",
      fetchKey,
    },
  );

  const { node: projectNode } = getFirstEdge(data?.salesforce?.uiapi?.query?.Acc_Project__c?.edges);
  const { node: pcrNode } = getFirstEdge(data?.salesforce?.uiapi?.query?.Acc_ProjectChangeRequest__c?.edges);
  const project = mapToProjectDto(projectNode, ["projectNumber", "status", "title", "roles", "competitionType"]);

  const partners = sortPartnersLeadFirst(
    mapToPartnerDtoArray(
      data?.salesforce?.uiapi?.query?.Acc_ProjectParticipant__c?.edges ?? [],
      ["id", "isLead", "isWithdrawn", "name"],
      {},
    ),
  );

  const pcrSpendProfile = mapPcrSpendProfileArray(data?.salesforce?.uiapi?.query?.Acc_IFSSpendProfile__c?.edges ?? [], [
    "capitalUsageType",
    "costCategoryId",
    "costOfEach",
    "costPerItem",
    "dateOtherFundingSecured",
    "daysSpentOnProject",
    "depreciationPeriod",
    "description",
    "grossCostOfRole",
    "id",
    "netPresentValue",
    "numberOfTimes",
    "overheadRate",
    "pcrItemId",
    "quantity",
    "ratePerDay",
    "residualValue",
    "subcontractorCountry",
    "subcontractorRoleAndDescription",
    "typeLabel",
    "utilisation",
    "value",
  ]);

  const documents = (pcrNode?.ContentDocumentLinks?.edges ?? []).map(node =>
    mapToDocumentSummaryDto(
      node,
      ["id", "dateCreated", "description", "fileName", "fileSize", "isOwner", "uploadedBy", "link", "linkedEntityId"],
      {
        type: "pcr",
        projectId,
        pcrId: pcrItemId,
      },
    ),
  );

  const costCategories = mapToCostCategoryDtoArray(data?.salesforce?.uiapi?.query?.Acc_CostCategory__c?.edges ?? [], [
    "id",
    "name",
    "displayOrder",
    "type",
    "competitionType",
    "organisationType",
  ]);

  const academicCostCategories = costCategories.filter(
    costCategory =>
      (costCategory.organisationType === PCROrganisationType.Academic &&
        costCategory.competitionType === project.competitionType) ||
      costCategory.type === CostCategoryType.Other_Public_Sector_Funding,
  );

  const pcrItem = mapPcrItemDto(
    pcrNode,
    [
      "awardRate",
      "contact1Email",
      "contact1Forename",
      "contact1Phone",
      "contact1Surname",
      "contact2Email",
      "contact2Forename",
      "contact2Phone",
      "contact2Surname",
      "financialYearEndDate",
      "financialYearEndTurnover",
      "hasOtherFunding",
      "isCommercialWork",
      "numberOfEmployees",
      "organisationName",
      "organisationType",
      "participantSize",
      "participantSizeLabel",
      "partnerType",
      "partnerTypeLabel",
      "projectCity",
      "projectLocation",
      "projectLocationLabel",
      "projectPostcode",
      "projectRole",
      "projectRoleLabel",
      "registeredAddress",
      "registrationNumber",
      "status",
      "tsbReference",
      "type",
    ],
    {},
  );

  const spendProfileCostCategories = costCategories.filter(
    x =>
      (x.competitionType === project.competitionType && x.organisationType === pcrItem.organisationType) ||
      x.type === CostCategoryType.Other_Public_Sector_Funding,
  );

  return {
    project,
    costCategories,
    academicCostCategories,
    spendProfileCostCategories,
    pcrItem,
    pcrSpendProfile,
    partners,
    documents,
    fragmentRef: data?.salesforce?.uiapi,
  };
};

export const getInitialAcademicCosts = (
  profile: PcrSpendProfileDto,
  costCategories: Pick<CostCategoryDto & { displayOrder: number }, "id" | "displayOrder" | "name" | "type">[],
) => {
  return costCategories
    .filter(x => x.type !== CostCategoryType.Other_Funding && x.type !== CostCategoryType.Other_Public_Sector_Funding)
    .sort((a, b) => (a.displayOrder > b.displayOrder ? 1 : -1))
    .map(x => {
      const matchingProfile = profile.costs.find(cost => cost.costCategoryId === x.id);
      if (matchingProfile) {
        return {
          id: matchingProfile.id,
          value: String(matchingProfile.value ?? 0),
          description: matchingProfile.description ?? "",
          costCategory: matchingProfile.costCategory,
          costCategoryId: matchingProfile.costCategoryId,
        };
      }

      return {
        id: "" as PcrId,
        value: "0",
        costCategoryId: x.id,
        description: x.name ?? "",
        costCategory: x.type,
      };
    });
};
