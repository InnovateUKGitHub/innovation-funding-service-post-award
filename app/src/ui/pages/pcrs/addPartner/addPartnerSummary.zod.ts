import { z } from "zod";
import { makeZodI18nMap } from "@shared/zodi18n";
import { PCROrganisationType, PCRProjectLocation, PCRProjectRole } from "@framework/constants/pcrConstants";
import { FormTypes } from "@ui/zod/FormTypes";
import { getTextValidation } from "@ui/zod/textareaValidator.zod";
import { evaluateObject } from "@ui/zod/helperValidators/helperValidators.zod";
import { getNumberValidation } from "@ui/zod/numericValidator.zod";

export const addPartnerErrorMap = makeZodI18nMap({ keyPrefix: ["pcr", "addPartner"] });

export const getAddPartnerSummarySchema = ({
  projectRole,
  organisationType,
}: {
  projectRole: PCRProjectRole;
  organisationType: PCROrganisationType;
}) =>
  evaluateObject(({ markedAsComplete: required }) => ({
    organisationName: getTextValidation({ required, maxLength: 256 }),
    registrationNumber: getTextValidation({
      required: required && organisationType === PCROrganisationType.Industrial,
      maxLength: 40,
    }),
    registeredAddress: getTextValidation({
      required: required && organisationType === PCROrganisationType.Industrial,
      maxLength: 32768,
    }),
    participantSize: required ? z.number().gt(0) : z.number().nullable().optional(),
    numberOfEmployees: getNumberValidation({
      min: 0,
      lt: 100_000_000,
      required: required && organisationType === PCROrganisationType.Industrial,
    }),
    financialYearEndDate:
      required && organisationType === PCROrganisationType.Industrial ? z.date() : z.date().nullable().optional(),
    financialYearEndTurnover: getNumberValidation({
      min: 0,
      required: required && organisationType === PCROrganisationType.Industrial,
    }),
    projectLocation: required
      ? z
          .number()
          .gt(0)
          .transform(x => x as PCRProjectLocation)
      : z
          .number()
          .transform(x => x as PCRProjectLocation)
          .nullable()
          .optional(),
    projectCity: getTextValidation({ required, maxLength: 40 }),
    projectPostcode: getTextValidation({ required: false, maxLength: 10 }),
    contact1Forename: getTextValidation({ required, maxLength: 50 }),
    contact1Surname: getTextValidation({ required, maxLength: 50 }),
    contact1Phone: getTextValidation({ required, maxLength: 20 }),
    contact1Email: getTextValidation({ required, maxLength: 255 }),
    contact2Forename: getTextValidation({
      required: required && projectRole === PCRProjectRole.ProjectLead,
      maxLength: 50,
    }),
    contact2Phone: getTextValidation({
      required: required && projectRole === PCRProjectRole.ProjectLead,
      maxLength: 20,
    }),
    contact2Surname: getTextValidation({
      required: required && projectRole === PCRProjectRole.ProjectLead,
      maxLength: 255,
    }),
    contact2Email: getTextValidation({
      required: required && projectRole === PCRProjectRole.ProjectLead,
      maxLength: 50,
    }),
    awardRate: getNumberValidation({ min: 0, max: 100, required }),
    partnerType: z.number().gt(0),
    projectRole: z.number().gt(0),
    isCommercialWork: required ? z.boolean() : z.boolean().nullable().optional(),
    hasOtherFunding: required ? z.boolean() : z.boolean().nullable().optional(),
    organisationType: getTextValidation({
      required: true,
      maxLength: 255,
    }),
    tsbReference: getTextValidation({
      required: required && organisationType === PCROrganisationType.Academic,
      maxLength: 255,
    }),
    markedAsComplete: z.literal(required),
    form: z.literal(FormTypes.PcrAddPartnerSummary),
  }));

export type AddPartnerSchemaType = ReturnType<typeof getAddPartnerSummarySchema>;
export type AddPartnerSchema = Omit<z.output<AddPartnerSchemaType>, "markedAsComplete"> & {
  markedAsComplete: boolean;
};
