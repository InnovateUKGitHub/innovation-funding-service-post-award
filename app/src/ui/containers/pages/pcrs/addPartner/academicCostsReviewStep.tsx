import React from "react";
import { CostCategoryDto } from "@framework/dtos/costCategoryDto";
import { PcrStepProps } from "@ui/containers/pages/pcrs/pcrWorkflow";
import { PCROrganisationType } from "@framework/constants/pcrConstants";
import { PCRSpendProfileAcademicCostDto } from "@framework/dtos/pcrSpendProfileDto";
import { sumBy } from "@framework/util/numberHelper";
import { Pending } from "@shared/pending";
import { Content } from "@ui/components/atomicDesign/molecules/Content/content";
import { Section } from "@ui/components/atomicDesign/molecules/Section/section";
import { Currency } from "@ui/components/atomicDesign/atoms/Currency/currency";
import { SimpleString } from "@ui/components/atomicDesign/atoms/SimpleString/simpleString";
import { createTypedTable } from "@ui/components/atomicDesign/molecules/Table/Table";
import { MountedHoc } from "@ui/components/atomicDesign/atoms/providers/Mounted/Mounted";
import { useStores } from "@ui/redux/storesProvider";
import { PCRPartnerAdditionItemDtoValidator } from "@ui/validation/validators/pcrDtoValidator";
import { PCRItemForPartnerAdditionDto } from "@framework/dtos/pcrDtos";
import { Loader } from "@ui/components/bjss/loading";
import { Link } from "@ui/components/atomicDesign/atoms/Links/links";

interface ContainerProps {
  costCategories: CostCategoryDto[];
}

interface Data {
  costCategory: CostCategoryDto;
  costDto: PCRSpendProfileAcademicCostDto;
}

const Table = createTypedTable<Data>();

class Component extends React.Component<
  PcrStepProps<PCRItemForPartnerAdditionDto, PCRPartnerAdditionItemDtoValidator> & ContainerProps,
  Data
> {
  render() {
    const { costCategories, pcrItem } = this.props;
    const data = costCategories
      .map(costCategory => {
        return {
          costCategory,
          costDto: this.props.pcrItem.spendProfile.costs.find(
            x => x.costCategoryId === costCategory.id,
          ) as PCRSpendProfileAcademicCostDto,
        };
      })
      .filter(x => !!x);
    const total = sumBy(data, x => (x.costDto ? x.costDto.value : 0) || 0);

    return (
      <MountedHoc>
        {state => (
          <Section title={x => x.pcrAddPartnerLabels.projectCostsHeading}>
            <Section title={x => x.pcrAddPartnerLabels.tsbReferenceHeading}>
              <SimpleString qa="tsbReference">{pcrItem.tsbReference}</SimpleString>
            </Section>

            <Section title={x => x.pages.pcrAddPartnerAcademicCosts.costsSectionTitle}>
              <Table.Table qa="costsTable" data={data}>
                <Table.String
                  header={x => x.pages.pcrAddPartnerAcademicCosts.categoryHeading}
                  qa="category"
                  value={x => x.costCategory.name}
                  footer={
                    state.isClient && (
                      <SimpleString className={"govuk-!-font-weight-bold"}>
                        <Content value={x => x.pages.pcrAddPartnerAcademicCosts.totalCosts} />
                      </SimpleString>
                    )
                  }
                />

                <Table.Currency
                  header={x => x.pages.pcrAddPartnerAcademicCosts.costHeading}
                  qa="cost"
                  value={x => (x.costDto ? x.costDto.value : 0)}
                  width={30}
                  footer={state.isClient && <Currency value={total} />}
                />
              </Table.Table>
            </Section>

            <Link
              styling="SecondaryButton"
              route={this.props.routes.pcrReviewItem.getLink({
                itemId: this.props.pcrItem.id,
                pcrId: this.props.pcr.id,
                projectId: this.props.project.id,
              })}
            >
              Return to summary
            </Link>
          </Section>
        )}
      </MountedHoc>
    );
  }
}

export const AcademicCostsReviewStep = (
  props: PcrStepProps<PCRItemForPartnerAdditionDto, PCRPartnerAdditionItemDtoValidator>,
) => {
  const stores = useStores();

  const costCategories = stores.costCategories
    .getAllUnfiltered()
    .then(allCostCategories =>
      allCostCategories.filter(
        costCategory =>
          costCategory.organisationType === PCROrganisationType.Academic &&
          costCategory.competitionType === props.project.competitionType,
      ),
    );

  return <Loader pending={Pending.combine({ costCategories })} render={x => <Component {...props} {...x} />} />;
};
