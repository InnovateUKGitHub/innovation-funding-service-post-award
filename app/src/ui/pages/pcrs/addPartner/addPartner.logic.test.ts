import { getInitialAcademicCosts } from "./addPartner.logic";

describe("getInitialAcademicCosts", () => {
  const spendProfile = {
    pcrItemId: "a0GAd000000to0zMAA" as PcrItemId,
    costs: [
      {
        id: "a08Ad00000BWIzzIAH" as CostId,
        costCategoryId: "a0626000007qoT1AAI" as CostCategoryId,
        value: 10000,
        description: "Directly incurred - Staff",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ00IAH" as CostId,
        costCategoryId: "a0626000007qoT2AAI" as CostCategoryId,
        value: 5000,
        description: "Directly incurred - Travel and subsistence",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ01IAH" as CostId,
        costCategoryId: "a0626000007qoT3AAI" as CostCategoryId,
        value: 2000,
        description: "Directly incurred - Equipment",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ02IAH" as CostId,
        costCategoryId: "a0626000007qoT4AAI" as CostCategoryId,
        value: 1000,
        description: "Directly incurred - Other costs",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ03IAH" as CostId,
        costCategoryId: "a0626000007qoT5AAI" as CostCategoryId,
        value: 500,
        description: "Directly allocated - Investigations",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ04IAH" as CostId,
        costCategoryId: "a0626000007qoT6AAI" as CostCategoryId,
        value: 0,
        description: "Directly allocated - Estates costs",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ05IAH" as CostId,
        costCategoryId: "a0626000007qoT7AAI" as CostCategoryId,
        value: 0,
        description: "Directly allocated - Other costs",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ06IAH" as CostId,
        costCategoryId: "a0626000007qoQpAAI" as CostCategoryId,
        value: 0,
        description: "Indirect costs - Investigations",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ07IAH" as CostId,
        costCategoryId: "a0626000007qoQqAAI" as CostCategoryId,
        value: 0,
        description: "Exceptions - Staff",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ08IAH" as CostId,
        costCategoryId: "a0626000007qoQrAAI" as CostCategoryId,
        value: 0,
        description: "Exceptions - Travel and subsistence",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ09IAH" as CostId,
        costCategoryId: "a0626000007qoQsAAI" as CostCategoryId,
        value: 0,
        description: "Exceptions - Equipment",
        costCategory: 2,
      },
      {
        id: "a08Ad00000BWJ0AIAX" as CostId,
        costCategoryId: "a0626000007qoQtAAI" as CostCategoryId,
        value: 0,
        description: "Exceptions - Other costs",
        costCategory: 2,
      },
    ],
    funds: [],
  };

  const academicCostCategories = [
    {
      id: "a0626000007qoT1AAI" as CostCategoryId,
      name: "Directly incurred - Staff",
      displayOrder: 8,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoT2AAI" as CostCategoryId,
      name: "Directly incurred - Travel and subsistence",
      displayOrder: 9,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoT3AAI" as CostCategoryId,
      name: "Directly incurred - Equipment",
      displayOrder: 10,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoT4AAI" as CostCategoryId,
      name: "Directly incurred - Other costs",
      displayOrder: 11,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoT5AAI" as CostCategoryId,
      name: "Directly allocated - Investigations",
      displayOrder: 12,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoT6AAI" as CostCategoryId,
      name: "Directly allocated - Estates costs",
      displayOrder: 13,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoT7AAI" as CostCategoryId,
      name: "Directly allocated - Other costs",
      displayOrder: 14,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoQpAAI" as CostCategoryId,
      name: "Indirect costs - Investigations",
      displayOrder: 15,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoQqAAI" as CostCategoryId,
      name: "Exceptions - Staff",
      displayOrder: 16,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoQrAAI" as CostCategoryId,
      name: "Exceptions - Travel and subsistence",
      displayOrder: 17,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoQsAAI" as CostCategoryId,
      name: "Exceptions - Equipment",
      displayOrder: 18,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoQtAAI" as CostCategoryId,
      name: "Exceptions - Other costs",
      displayOrder: 19,
      type: 2,
      competitionType: "CR&D",
      organisationType: "Academic",
    },
    {
      id: "a0626000007qoRSAAY" as CostCategoryId,
      name: "Other public sector funding",
      displayOrder: 91,
      type: 11,
      competitionType: "unknown",
      organisationType: "unknown",
    },
  ];

  const emptySpendProfile = {
    pcrItemId: "a0GAd00000118e5MAA" as PcrItemId,
    costs: [],
    funds: [],
  };

  it("should return the formatted initial academic costs without other public sector funding if there are already costs saved", () => {
    expect(getInitialAcademicCosts(spendProfile, academicCostCategories)).toMatchSnapshot();
  });

  it("should initialise an empty spend profile with default values", () => {
    expect(getInitialAcademicCosts(emptySpendProfile, academicCostCategories)).toMatchSnapshot();
  });

  it("should sort the costs accurately even if the data is not consistently ordered", () => {
    const initialCosts = getInitialAcademicCosts(spendProfile, academicCostCategories);
    const initialCostsWithRandomOrder = getInitialAcademicCosts(
      spendProfile,
      academicCostCategories.sort(() => Math.random() - 0.5),
    );

    expect(initialCosts).toEqual(initialCostsWithRandomOrder);
  });
});
