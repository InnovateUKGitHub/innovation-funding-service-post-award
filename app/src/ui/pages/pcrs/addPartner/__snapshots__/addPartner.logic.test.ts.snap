// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getInitialAcademicCosts should initialise an empty spend profile with default values 1`] = `
[
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT1AAI",
    "description": "Directly incurred - Staff",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT2AAI",
    "description": "Directly incurred - Travel and subsistence",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT3AAI",
    "description": "Directly incurred - Equipment",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT4AAI",
    "description": "Directly incurred - Other costs",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT5AAI",
    "description": "Directly allocated - Investigations",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT6AAI",
    "description": "Directly allocated - Estates costs",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT7AAI",
    "description": "Directly allocated - Other costs",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoQpAAI",
    "description": "Indirect costs - Investigations",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoQqAAI",
    "description": "Exceptions - Staff",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoQrAAI",
    "description": "Exceptions - Travel and subsistence",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoQsAAI",
    "description": "Exceptions - Equipment",
    "id": "",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoQtAAI",
    "description": "Exceptions - Other costs",
    "id": "",
    "value": "0",
  },
]
`;

exports[`getInitialAcademicCosts should return the formatted initial academic costs without other public sector funding if there are already costs saved 1`] = `
[
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT1AAI",
    "description": "Directly incurred - Staff",
    "id": "a08Ad00000BWIzzIAH",
    "value": "10000",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT2AAI",
    "description": "Directly incurred - Travel and subsistence",
    "id": "a08Ad00000BWJ00IAH",
    "value": "5000",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT3AAI",
    "description": "Directly incurred - Equipment",
    "id": "a08Ad00000BWJ01IAH",
    "value": "2000",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT4AAI",
    "description": "Directly incurred - Other costs",
    "id": "a08Ad00000BWJ02IAH",
    "value": "1000",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT5AAI",
    "description": "Directly allocated - Investigations",
    "id": "a08Ad00000BWJ03IAH",
    "value": "500",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT6AAI",
    "description": "Directly allocated - Estates costs",
    "id": "a08Ad00000BWJ04IAH",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoT7AAI",
    "description": "Directly allocated - Other costs",
    "id": "a08Ad00000BWJ05IAH",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoQpAAI",
    "description": "Indirect costs - Investigations",
    "id": "a08Ad00000BWJ06IAH",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoQqAAI",
    "description": "Exceptions - Staff",
    "id": "a08Ad00000BWJ07IAH",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoQrAAI",
    "description": "Exceptions - Travel and subsistence",
    "id": "a08Ad00000BWJ08IAH",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoQsAAI",
    "description": "Exceptions - Equipment",
    "id": "a08Ad00000BWJ09IAH",
    "value": "0",
  },
  {
    "costCategory": 2,
    "costCategoryId": "a0626000007qoQtAAI",
    "description": "Exceptions - Other costs",
    "id": "a08Ad00000BWJ0AIAX",
    "value": "0",
  },
]
`;
